core/src/test/java/ai/timefold/solver/core/impl/domain/variable/cascade/CollectionCascadingUpdateShadowVariableListenerTest.java

```diff
@@ -0,0 +1,171 @@
+package ai.timefold.solver.core.impl.domain.variable.cascade;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+
+import ai.timefold.solver.core.api.score.buildin.simple.SimpleScore;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
+import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
+import ai.timefold.solver.core.impl.testdata.domain.cascade.multiple_var.supply.TestdataMultipleCascadingWithSupplyEntity;
+import ai.timefold.solver.core.impl.testdata.domain.cascade.multiple_var.supply.TestdataMultipleCascadingWithSupplySolution;
+import ai.timefold.solver.core.impl.testdata.domain.cascade.multiple_var.supply.TestdataMultipleCascadingWithSupplyValue;
+import ai.timefold.solver.core.impl.testdata.util.PlannerTestUtils;
+
+import org.junit.jupiter.api.Test;
+
+class CollectionCascadingUpdateShadowVariableListenerTest {
```

Comment:
Arguably, if the test can be parameterized, we should be able to do this with 1 test class, not 4, right? The test classes are nearly identical.